### PR TITLE
Add support for on_table_exists in table materialization

### PR DESCRIPTION
**Pull request by hovaesco**
## Overview

- Description: Add support for on_table_exists in table materialization
- Related issue(s): https://github.com/starburstdata/dbt-trino/issues/26
- Related code pull request(s):
- Related link(s):

## Checklist

- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] `README.md` updated and added information about my change
- [x] `CHANGELOG.md` updated and added information about my change

### CODE REVIEW

File: dbt/include/trino/macros/materializations/table.sql [hunk 1]
```diff
@@ -1,58 +1,27 @@
...
   {%- set identifier = model['alias'] -%}
-  {%- set tmp_identifier = model['name'] + '__dbt_tmp' -%}
-  {%- set backup_identifier = model['name'] + '__dbt_backup' -%}
 
   {%- set old_relation = adapter.get_relation(database=database, schema=schema, identifier=identifier) -%}
   {%- set target_relation = api.Relation.create(identifier=identifier,
                                                 schema=schema,
                                                 database=database,
                                                 type='table') -%}
-  {%- set intermediate_relation = api.Relation.create(identifier=tmp_identifier,
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table.sql:11]: Can you please explain in the commit comment why it was considered that the intermediate relation was superfluos?

**Comment by leniartek** [comment]
LGTM  👍

File: dbt/include/trino/macros/materializations/table.sql [hunk 2]
```diff
@@ -1,58 +1,27 @@
...
-
-  -- `COMMIT` happens here
-  {{ adapter.commit() }}
-
-  -- finally, drop the existing/backup relation after the commit
-  {{ drop_relation_if_exists(backup_relation) }}
+  {{ run_hooks(post_hooks) }}
 
-  {{ run_hooks(post_hooks, inside_transaction=False) }}
+  {% do persist_docs(target_relation, model) %}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table.sql:24]: what is this line needed for?

File: dbt/include/trino/macros/materializations/table.sql [hunk 3]
```diff
@@ -1,58 +1,27 @@
...
-
-
-  -- drop the temp relations if they exists for some reason
-  {{ adapter.drop_relation(intermediate_relation) }}
-  {{ adapter.drop_relation(backup_relation) }}
-
-  {{ run_hooks(pre_hooks, inside_transaction=False) }}
+  {{ run_hooks(pre_hooks) }}
 
-  -- `BEGIN` happens here:
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table.sql:32]: I'm fine with dropping the transaction logic because Trino doesn't support transactions anyway.

File: dbt/include/trino/macros/materializations/table.sql [hunk 4]
```diff
@@ -1,58 +1,27 @@
...
-    {{ create_table_as(False, intermediate_relation, sql) }}
+    {{ create_table_as(False, target_relation, sql) }}
   {%- endcall %}
 
-  -- cleanup
-  {% if old_relation is not none %}
-      {{ adapter.rename_relation(old_relation, backup_relation) }}
-  {% endif %}
-
-  {{ adapter.rename_relation(intermediate_relation, target_relation) }}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table.sql:45]: The main advantage of doing the  intermediate relation renaming is that the `table` materialization appears to be atomic to the end user.

The change proposed in this PR however handles the `table` materialization in two steps:
- the first one drops the existing table
- the second one creates the table which corresponds to the newest model

Now if the execution flow is interrupted in between, the existing table  would be dropped and there would be no backup available.

I do know that AWS Glue doesn't support table renames and that dbt-trino would need to cope with this aspect. However we need to keep in mind that dbt-trino is a generic (and not AWS Glue specific) library.

Let's investigate a bit more how we could cover the need of dealing with AWS Glue tables in dbt-trino.

Comment (findinpath) [reply dbt/include/trino/macros/materializations/table.sql:45]: We could verify whether table renames are supported (create a table , try to rename it and see whether an exception with `NOT_SUPPORTED`  error code is being returned) and fallback to this rather simplistic flow. However this is a rather big overhead because most of the databases actually do support table renames.

What about providing a specific materialization for aws-glue tables? This would be a reasonable compromise right?

**Review by findinpath — CHANGES_REQUESTED** [review]
Please consider providing a specific materialization for AWS Glue tables instead of degrading the existing functionality of the `table` materialization

**Comment by hovaesco** [comment]
I've added a new materialization `table_glue` which is compatible with AWS Glue which does not support table renames. `table` materialization is refactored and redundant logic has been deleted.

**Comment by mdesmet** [comment]
LGTM.

As mentioned on Slack, I think we should find a way to support renaming Glue tables.

**Comment by hovaesco** [comment]
For completeness of dbt-trino I complete agree that we need to find a way to support it in Trino. 

Other adapters have drop and recreate table materialization set as a default one for tables. I'm considering calling a new one `table_drop` to indicate that is dropping and recreating a table under the hood and describe in README that it supports Glue. 

Users gain optionality to choose which built-in table macro mode they do need depending on various needs.

File: dbt/include/trino/macros/materializations/table_glue.sql [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{% materialization table_glue, adapter = 'trino' %}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table_glue.sql:1]: Looks good.

File: dbt/include/trino/macros/materializations/table.sql [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{% materialization table, adapter='trino' %}
+{% materialization table, adapter = 'trino' %}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table.sql:1]: Please cleanup the unnecessary changes from this file.

File: dbt/include/trino/macros/materializations/table.sql [hunk 7]
```diff
@@ -8,51 +9,58 @@
...
 
-  {%- set backup_relation_type = 'table' if old_relation is none else old_relation.type -%}
-  {%- set backup_relation = api.Relation.create(identifier=backup_identifier,
-                                                schema=schema,
-                                                database=database,
-                                                type=backup_relation_type) -%}
-
-  {%- set exists_as_table = (old_relation is not none and old_relation.is_table) -%}
-  {%- set exists_as_view = (old_relation is not none and old_relation.is_view) -%}
+  {% if table_exists_strategy == 'RENAME' %}
```
Comment (findinpath) [inline dbt/include/trino/macros/materializations/table.sql:13]: Check whether the constant used for determining the handling for table existence is valid.
Use https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/include/global_project/macros/materializations/models/incremental/on_schema_change.sql as template.

File: README.md [hunk 8]
```diff
@@ -180,6 +180,43 @@ NOTE that this functionality works on incremental models that use partitioning:
...
+or in `dbt_project.yaml`:
+
+```yaml
+models:
+  path:
+    materialized: table
+    +table_exists_strategy: DROP
+```
+
+Using `table` materialization and `table_exists_strategy = 'RENAME'` with AWS Glue may result in below error:
```
Comment (findinpath) [inline README.md:214]: `table_exists_strategy` -> `on_table_exists` ?

This seems in line with https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/include/global_project/macros/materializations/models/incremental/on_schema_change.sql 

which is used for the `incremental` materialization in dbt core.

Comment (hovaesco) [reply README.md]: Done

File: dbt/include/trino/macros/materializations/table.sql (hunk 7 above)
Comment (findinpath) [reply dbt/include/trino/macros/materializations/table.sql]: Please see https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/include/global_project/macros/materializations/models/incremental/on_schema_change.sql#L3

File: docker/dbt/jaffle_shop/dbt_project.yml [hunk 9]
```diff
@@ -22,6 +22,9 @@ models:
     materialized: table
     staging:
       materialized: view
+    table_drop:
+      materialized: table
+      +on_table_exists: dupa
```
Comment (findinpath) [inline docker/dbt/jaffle_shop/dbt_project.yml:27]: I assume this is an unintended change.

**Review by findinpath — APPROVED** [review]

File: docker/dbt/jaffle_shop/dbt_project.yml (hunk 9 above)
Comment (hovaesco) [reply docker/dbt/jaffle_shop/dbt_project.yml]: thanks, that's a not polite word usually used to test things in Polish :D

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,12 +1,14 @@
 ## dbt-trino TBD (TBD)
 
 ### Features
+- Add support for on_table_exists in table materialization ([#26](https://github.com/starburstdata/dbt-trino/issues/26), [#54](https://github.com/starburstdata/dbt-trino/pull/54))
```
Comment (findinpath) [inline CHANGELOG.md:4]: nit : `on_table_exists`

**Review by findinpath — APPROVED** [review]
